console/src/components/Services/Data/Add/AddTable.js

```diff
@@ -124,137 +145,244 @@ class AddTable extends Component {
+  // and the second arg is ignored
+  checkAndDispatch(validated, notificationArray) {
+    if (validated === true || validated === '') return true;
+    else if (validated === false) {
       this.props.dispatch(
         showErrorNotification(
-          'Error creating table!',
-          'Select atleast one primary key',
-          {
-            custom: ATLEAST_ONE_PRIMARY_KEY_MSG,
-          }
+          notificationArray[0],
+          notificationArray[1],
+          notificationArray[2],
+          notificationArray[3]
         )
       );
       return false;
-    }
-    return true;
-  }
-
-  tableNameCheck() {
-    if (this.props.tableName === '' || this.props.tableName === null) {
-      // this.props.dispatch(validationError('Table name cannot be empty'));
-      // alert('Table name cannot be empty');
+    } else if (typeof validated === 'string') {
       this.props.dispatch(
         showErrorNotification(
-          'Error creating table!',
-          'Table name cannot be empty',
-          {
-            custom: 'Table name cannot be empty. Please add a name',
-          }
+          notificationArray[0],
+          validated,
+          notificationArray[2],
+          { custom: validated }
         )
       );
       return false;
-    } else if (!gqlPattern.test(this.props.tableName)) {
+    } else if (typeof validated === 'object') {
       this.props.dispatch(
         showErrorNotification(
-          gqlTableErrorNotif[0],
-          gqlTableErrorNotif[1],
-          gqlTableErrorNotif[3]
+          notificationArray[0],
+          notificationArray[1],
+          notificationArray[2],
+          notificationArray[3]
         )
       );
       return false;
     }
+  }
+
+  tableNameCheck() {
+    return gqlPattern.test(this.props.tableName);
+  }
+
+  isModified(x) {
+    if (x === undefined) return false;
+    else if (typeof x === 'string' && /^..*$/.test(x)) return true;
+
+    return false;
+  }
+
+  // return a shallow copy of a columns array, trimming those columns with
+  // no non-default values from the end. Invalids in the middle retained
+  trimEmptyColumns(Columns) {
+    let c = Columns.slice(0);
+
+    for (let i = c.length - 1; i >= 0; i--) {
+      if (
+        this.isModified(c[i].name) ||
+        this.isModified(c[i].type) ||
+        this.isModified(c[i].default)
+      ) {
+        return c;
+      }
+      c = c.slice(0, c.length - 1);
+    }
+    return c;
+  }
+
+  validateEnoughColumns(cols) {
+    return cols.length > 0;
+  }
+
+  validateColumnNames(cols) {
+    const l = cols.length;
+
+    for (let i = 0; i < l; i++) {
+      if (!gqlPattern.test(cols[i].name)) {
+        return (
+          'Invalid name for column ' +
+          i.toString() +
+          ', ' +
+          cols[i].name.toString() +
+          ' is Invalid. ' +
+          'Column names must be letters, numbers, and _ and start with a letter'
+        );
+      }
+    }
+    return '';
+  }
+
+  isValidType(s) {
```

Comment:
Just noticed that this prevents creating columns of non-common types. Considering the type selector is a dropdown with all allowed values, we can probably skip this check.

All else looks good to merge.

Reply:
fixed